src/function/cast/decimal_cast.cpp

```diff
@@ -114,11 +114,41 @@ struct DecimalScaleDownOperator {
 	}
 };
 
+// This function detects if we can scale a decimal down to another.
+template <class INPUT_TYPE>
+bool CanScaleDownDecimal(INPUT_TYPE input, DecimalScaleInput<INPUT_TYPE> &data) {
+	int64_t divisor = static_cast<int64_t>(pow(10, data.source_scale));
+	auto value = input % divisor;
+	auto rounded_input = input;
+	if (rounded_input < 0) {
+		rounded_input *= -1;
+		value *= -1;
+	}
+	if (value >= divisor / 2) {
+		rounded_input += divisor;
+	}
+	return rounded_input < data.limit && rounded_input > -data.limit;
+}
+
+bool CanScaleDownDecimal(hugeint_t input, DecimalScaleInput<hugeint_t> &data) {
+	auto divisor = Hugeint::Pow(10, data.source_scale);
```

Comment:
Can we use `Hugeint::POWERS_OF_TEN` instead here? That avoids having to do this `while` loop for every value we convert